src/Common/Core/Impl/Threading/AsyncReaderWriterLock.Queue.cs

```diff
@@ -134,7 +134,7 @@ private class Queue {
             public LockSource[] Remove(LockSource item) {
                 lock (this) {
                     var next = item.Next;
-                    Interlocked.CompareExchange(ref _head, next, item);
+                    var oldHead = Interlocked.CompareExchange(ref _head, next, item);
```

Comment:
That was my original idea, but it happened that it is not enough - locks around several operations with the list are still required. There is an existing `LinkedList<T>`, but it is slower and creates tons of garbage in memory.